motion_profiles.py

```diff
@@ -0,0 +1,101 @@
+        # Same slopes means parallel or collinear
+        if line1[0] == line2[0]:
+            raise ValueError("Lines never intersect at a single point")
+
+        x_intersect = (line2[1] - line1[1]) / (line1[0] - line2[0])
+        y_intersect = line1[0] * x_intersect + line1[1]
+
+        return (x_intersect, y_intersect)
+
+    @staticmethod
+    def generate_motion_profile(acceleration_time: float, deceleration_time: float,
+                                max_speed: float, target_distance: float):
+        """Create a motion profile to efficiently travel `target_distance`
+         using the specified parameters
+
+        `acceleration_time`: Time it takes for robot to accelerate from rest to maximum speed.
+        `deceleration_time`: Time it takes for robot to decelerate from maximum speed to rest.
+        `max_speed`: The maximum speed the robot can go.
+        `target_distance`: The distance the robot should travel using the motion profile.
+
+        ** Note: The units don't matter, as long as they are all the same - but anyone**
+        **caught using non-metric units will be publicly shamed.**
+        """
+        acceleration_distance = 0.5 * acceleration_time * max_speed
+        acceleration = max_speed / acceleration_time
+        deceleration_distance = 0.5 * deceleration_time * max_speed
+        deceleration = max_speed / deceleration_time
+
+        overshoot_distance = (acceleration_distance +
+                              deceleration_distance) - target_distance
+        overshoot_time = overshoot_distance / max_speed
+        end_time = acceleration_time + deceleration_time - overshoot_time
+
+        # Robot should spend some time at max speed
+        if overshoot_distance <= 0:
+            motion_profile = MotionProfile(
+                acceleration, deceleration, acceleration_time,
+                deceleration_time, max_speed, end_time)
+            return motion_profile
+
+        # If robot accelerates all the way to max speed, it will overshoot target
+        else:
+            acceleration_line = (acceleration, 0)
+            deceleration_line = (deceleration, -end_time * deceleration)
+            max_speed_point = MotionProfile.intersection(
+                acceleration_line, deceleration_line)
+            max_speed_time = max_speed_point[0]
+            motion_profile = MotionProfile(
+                acceleration, deceleration, max_speed_time, max_speed_time, max_speed, end_time)
+            return motion_profile
+
+    def position(self, time):
+        """Get the optimal position at a specific time"""
+        def triangle_area(slope, intercept, time):
+            """Find the area under a line described by y = `slope` *
+            x + `intercept` for the interval [`0`, `time`]
+
+            This is equivalent to the area of a right-angled
+            triangle with base `time` and hypotenuse described by the line"""
+            return 0.5 * time * ((slope * time) + intercept)
+
```

Comment:
Inner functions can be a great way to abstract complex features, however, this visually looks like a mistake 😉

Since we can't (I don't think) see the document associated with the `triangle_area` function, perhaps it would look better if it was a comment before it, and explains that it is an inner, helper function?